### PR TITLE
refactor: add helper class for validating inputs

### LINKED ISSUE #634: [SPK] create helper methods for validating cli input
**As a:**
Developer

**I want:**
to be able to call helper that can validate input to check for empty/null/expression validity.  

**So that:**

**Describe the solution you'd like:**
helper methods 

**Describe alternatives you've considered:**

**Additional context:**
adopted these methods from an existing project https://github.com/serverless/serverless-azure-functions/blob/master/src/shared/guard.ts

**Does this require updates to documentation?:**

### PR DESCRIPTION
**Pull request by mydiemho**
fixes https://github.com/microsoft/bedrock/issues/634

### CODE REVIEW

File: src/lib/guard.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  message?: string
+) => {
+  if (!!value === false) {
+    message =
+      message || `'${paramName || "value"}' cannot be null or undefined`;
+    throw new Error(message);
+  }
+};
+
+export const guardTypeOf = (value: any, type: string, message?: string) => {
```
Comment (mtarng) [inline src/lib/guard.ts:43]: JSDocs, otherwise LGTM.

**Review by mtarng — APPROVED** [review]

**Comment by bnookala** [comment]
I think you should make clear that this should be using for coercing and validating user inputs - as opposed to replacing typescript's type guards etc, but otherwise it looks ok

**Review by bnookala — APPROVED** [review]

File: src/lib/guard.ts (hunk 1 above)
Comment (evanlouie) [inline src/lib/guard.ts:43]: `value` needs be of type `number | string | boolean`.  `typeof` can't be used on `Array` or `Object` (`Array` extends `Object`, so `typeof` cant be confidently used).

If you want to support `number | string | boolean | Array | Object`, your code would have to look like:

```ts
export const guardTypeOf  = <T>(
  value: T,
  type: "number" | "string" | "boolean" | "array" | "object",
  message?: string
) => {
  if (
    ["number", "string", "boolean", "object"].includes(type) &&
    typeof value !== type
  ) {
    message =
      message || `${value} must be of type ${type}, ${typeof value} given`;
    throw new Error(message);
  } else if (type === "array" && Array.isArray(value) === false) {
    message =
      message || `${value} must be of type Array, ${typeof value} given`;
    throw new Error(message);
  }
};

```

**Review by evanlouie — CHANGES_REQUESTED** [review]

File: src/commands/service/create.ts [hunk 2]
```diff
@@ -41,26 +42,12 @@ export const createCommandDecorator = (command: commander.Command): void => {
...
-        if (typeof maintainerEmail !== "string") {
-          throw new Error(
-            `maintainerEmail must be of type 'string', ${typeof maintainerEmail} given.`
-          );
-        }
+        const stringType = "string";
+        guardTypeOf(serviceName, stringType);
+        guardTypeOf(packagesDir, stringType);
+        guardTypeOf(maintainerName, stringType);
+        guardTypeOf(maintainerEmail, stringType);
```
Comment (evanlouie) [inline src/commands/service/create.ts:49]: Does `guardTypeOf` actually act as a type guard for any of the variables?
Following lines 49, does the compiler actually know `serviceName`, `packagesDir`, etc... are of type `string`?
I don't believe the current setup will actually allow the compiler to know the types of the variables (ie; their still `any`) after the `guardTypeOf` calls; whereas the compiler knows their of type `string` using the `if (typeof maintainerEmail !== "string") {...` syntax.

Comment (evanlouie) [reply src/commands/service/create.ts:49]: Just played with this a bit, just as I feared

Fails:

```ts
const log = (msg: string) => console.log(msg);
const foo: unknown = "i'm a string";
guardTypeOf(foo, "string");
log(foo);
```

Passes:

```ts
const log = (msg: string) => console.log(msg);
const foo: unknown = "i'm a string";
if (typeof foo !== "string") {
  throw new Error("Expected string");
}
log(foo);
```

We should maintain to just use vanilla TS typeguards for now (`if (typeof ...`)).

Comment (mydiemho) [reply src/commands/service/create.ts:49]: I'm ok with using vanilla guards.  But is there a way to abstract it out to keep the main logic simple?

Comment (evanlouie) [reply src/commands/service/create.ts:49]: Sadly not yet. TS is planning on supporting `assert`ion guards which do exactly this in TS 3.7 https://devblogs.microsoft.com/typescript/announcing-typescript-3-7-beta/

But until then, the compiler isn't smart enough to really support a clean abstraction :(

**Pull request closed without merging**